File: backend/user-order-service/src/models/Delivery.js
/**
 * Delivery Model
 */

const mongoose = require('mongoose');
const { DELIVERY_STATUS } = require('@medassist/shared/constants');

const locationSchema = new mongoose.Schema({
  lat: { type: Number, required: true },
  lon: { type: Number, required: true }
}, { _id: false });

const deliverySchema = new mongoose.Schema({
  order_id: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'Order', 
    required: true,
    index: true
  },
  driver_id: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'User',
    index: true
  },
  assigned_at: { 
    type: Date 
  },
  pickup_at: { 
    type: Date 
  },
  delivered_at: { 
    type: Date 
  },
  status: { 
    type: String, 
    enum: Object.values(DELIVERY_STATUS),
    default: DELIVERY_STATUS.ASSIGNED,
    index: true
  },
  current_location: locationSchema,
  pickup_location: locationSchema,
  delivery_location: locationSchema,
  estimated_delivery_time: {
    type: Date
  },
  notes: {
    type: String,
    trim: true
  }
}, { 
  timestamps: true 
});

// Indexes
deliverySchema.index({ driver_id: 1, status: 1 });
deliverySchema.index({ assigned_at: -1 });

module.exports = mongoose.model('Delivery', deliverySchema);

File: backend/driver-service/app/models/delivery.py
"""
Delivery models for Driver Service
"""

from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from enum import Enum


class DeliveryStatus(str, Enum):
    ASSIGNED = "assigned"
    PICKED_UP = "picked_up"
    IN_TRANSIT = "in_transit"
    DELIVERED = "delivered"
    FAILED = "failed"


class Location(BaseModel):
    lat: float = Field(ge=-90, le=90)
    lon: float = Field(ge=-180, le=180)


class DeliveryStatusUpdate(BaseModel):
    status: DeliveryStatus
    lat: Optional[float] = Field(default=None, ge=-90, le=90)
    lon: Optional[float] = Field(default=None, ge=-180, le=180)
    notes: Optional[str] = None


class DeliveryOTPConfirm(BaseModel):
    otp: str = Field(pattern=r'^\d{6}$')


class DeliveryResponse(BaseModel):
    id: str
    order_id: str
    driver_id: Optional[str] = None
    status: DeliveryStatus
    assigned_at: Optional[datetime] = None
    pickup_at: Optional[datetime] = None
    delivered_at: Optional[datetime] = None
    current_location: Optional[dict] = None
    
    class Config:
        from_attributes = True
